go/types/subtype.go

```diff
@@ -126,3 +126,110 @@ func compoundSubtype(requiredType, concreteType *Type, parentStructTypes []*Type
 	}
 	return isSubtype(requiredType, concreteType, parentStructTypes)
 }
+
+// IsValueSubtypeOf returns whether a value is a subtype of a type.
+func IsValueSubtypeOf(v Value, t *Type) bool {
+	return isValueSubtypeOf(v, t, nil)
+}
+
+func isValueSubtypeOf(v Value, t *Type, parentStructTypes []*Type) bool {
+	switch t.TargetKind() {
+	case BoolKind, NumberKind, StringKind, BlobKind, TypeKind:
+		return v.Kind() == t.TargetKind()
+	case ValueKind:
+		return true
+	case UnionKind:
+		for _, et := range t.Desc.(CompoundDesc).ElemTypes {
+			if IsValueSubtypeOf(v, et) {
+				return true
+			}
+		}
+		return false
+	default:
+		if v.Kind() != t.TargetKind() {
```

Comment:
do we want to put a panic here if t.TargetKind() is CycleKind? Seems like that's a programmer error rather than a false.